### PR TITLE
feat: Add trace transaction API

### PR DESCRIPTION
**Pull request by snissn**
## Related Issues
This PR closes issue https://github.com/filecoin-project/lotus/issues/12068
### Proposed Changes
- Add json RPC support for [trace_transaction](https://openethereum.github.io/JSONRPC-trace-module#trace_transaction) to lotus

## Additional Info
This RPC API is an Ethereum standard feature for data access and is needed by Gnosis Safe Wallet Account Abstraction - https://docs.safe.global/core-api/api-safe-transaction-service/rpc-requirements#what-rpc-methods-are-used 

## Checklist

Before you mark the PR ready for review, please make sure that:

- [x] Commits have a clear commit message.
- [ ] PR title is in the form of of `<PR type>: <area>: <change being made>`
  - example: ` fix: mempool: Introduce a cache for valid signatures`
  - `PR type`: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test
  - `area`, e.g. api, chain, state, market, mempool, multisig, networking, paych, proving, sealing, wallet, deps
- [x] If the PR affects users (e.g., new feature, bug fix, system requirements change), update the CHANGELOG.md and add details to the UNRELEASED section.
- [ ] New features have usage guidelines and / or documentation updates in
  - [ ] [Lotus Documentation](https://lotus.filecoin.io)
  - [ ] [Discussion Tutorials](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [x] Tests exist for new functionality or change in behavior
- [x] CI is green

### Testing

- Use curl to verify the accuracy of the api endpoint

```bash
curl --request POST \
     --url http://localhost:1234/rpc/v1 \
     --header 'accept: application/json' \
     --header 'content-type: application/json' \
     --data '
{
  "id": 1,
  "jsonrpc": "2.0",
  "method": "trace_transaction",
  "params": [
    "0xc088d3808cbce1cc06019e42eb4d642e7ba51848282bf4735d518f36da74e036"
   ]
}
'




{"id":1,"jsonrpc":"2.0","result":[{"type":"call","subtraces":3,"traceAddress":[],"action":{"callType":"call","from":"0x349ebc0f74f6713940977e1883d29f35e4b12640","to":"0x51e1f72b655528de2d4d88e70bd53774db8d0b0c","gas":"0x464450a","value":"0x56364ccee568c3000","input":"0xb6b55f2500000000000000000000000000000000000000000000000563503668c2130000"},"result":{"gasUsed":"0x1241ea1","output":"0x0000000000000000000000000000000000000000000000056364c72c1f83b350"},"blockHash":"0x6edf16a97b9a0d6722ab36a1234b24020328a41dffc9ca380457c5b5346b2059","blockNumber":1679653,"transactionHash":"0xc088d3808cbce1cc06019e42eb4d642e7ba51848282bf4735d518f36da74e036","transactionPosition":1},{"type":"call","subtraces":0,"traceAddress":[0],"action":{"callType":"staticcall","from":"0x51e1f72b655528de2d4d88e70bd53774db8d0b0c","to":"0x493cd7e48854582eb9aa3a303c06984d23f5a30e","gas":"0x3f53386","value":"0x0","input":"0x18160ddd"},"result":{"gasUsed":"0x413340","output":"0x000000000000000000000000000000000000000000003c5ac8f816829eec6d7c"},"blockHash":"0x6edf16a97b9a0d6722ab36a1234b24020328a41dffc9ca380457c5b5346b2059","blockNumber":1679653,"transactionHash":"0xc088d3808cbce1cc06019e42eb4d642e7ba51848282bf4735d518f36da74e036","transactionPosition":1},{"type":"call","subtraces":0,"traceAddress":[1],"action":{"callType":"staticcall","from":"0x51e1f72b655528de2d4d88e70bd53774db8d0b0c","to":"0x9b038febc0b3c501fdad8b0adc4386cc21aadff4","gas":"0x39747e0","value":"0x0","input":"0x7da65e790000000000000000000000000000000000000000000000056364ccee568c300000000000000000000000000000000000000000000000000000000000000dbba000000000000000000000000000000000000000000000000000000000000f4240000000000000000000000000000000000000000000003c5ac8f816829eec6d7c000000000000000000000000000000000000000000003c5ac9389766fb4024cd0000000000000000000000000000000000000000000000000000000000000000"},"result":{"gasUsed":"0x27d581","output":"0x0000000000000000000000000000000000000000000000056364c72c1f83b350"},"blockHash":"0x6edf16a97b9a0d6722ab36a1234b24020328a41dffc9ca380457c5b5346b2059","blockNumber":1679653,"transactionHash":"0xc088d3808cbce1cc06019e42eb4d642e7ba51848282bf4735d518f36da74e036","transactionPosition":1},{"type":"call","subtraces":0,"traceAddress":[2],"action":{"callType":"call","from":"0x51e1f72b655528de2d4d88e70bd53774db8d0b0c","to":"0x493cd7e48854582eb9aa3a303c06984d23f5a30e","gas":"0x2d01398","value":"0x0","input":"0x40c10f19000000000000000000000000349ebc0f74f6713940977e1883d29f35e4b126400000000000000000000000000000000000000000000000056364c72c1f83b350"},"result":{"gasUsed":"0x1f2d492","output":"0x"},"blockHash":"0x6edf16a97b9a0d6722ab36a1234b24020328a41dffc9ca380457c5b5346b2059","blockNumber":1679653,"transactionHash":"0xc088d3808cbce1cc06019e42eb4d642e7ba51848282bf4735d518f36da74e036","transactionPosition":1}]}
```

```
% curl --request POST \
     --url http://localhost:1234/rpc/v1 \
     --header 'accept: application/json' \
     --header 'content-type: application/json' \
     --data '
{
  "id": 1,
  "jsonrpc": "2.0",
  "method": "trace_transaction",
  "params": [
    "0xc088d3808cbce1cc06019e42eb4d642e7ba51848282bf4735d518f36da74e033"
   ]
}
'




{"error":{"code":1,"message":"transaction not found"},"id":1,"jsonrpc":"2.0"}
```

### CODE REVIEW

File: node/impl/full/eth.go [hunk 1]
```diff
@@ -976,6 +977,46 @@ func (a *EthModule) EthTraceReplayBlockTransactions(ctx context.Context, blkNum
...
+	tx, err := a.EthGetTransactionByHash(ctx, &ethTxHash)
+	if err != nil {
+		return nil, xerrors.Errorf("cannot get transaction by hash: %w", err)
+	}
+
+	if tx == nil {
+		return nil, xerrors.Errorf("transaction not found")
+	}
+
+	blockTraces, err := a.EthTraceBlock(ctx, strconv.FormatUint(uint64(*tx.BlockNumber), 10))
```
Comment (aarshkshah1992) [inline node/impl/full/eth.go:997]: The `BlockNumber` will be nil if the transaction is still in the mpool/pending. We should just error out in this case as there's no trace for pending transactions and document the API to explicitly say so.

**Review by aarshkshah1992 — COMMENTED** [review]
Need to fix/decide on the handling of transactions that have yet to be mined.

**Comment by aarshkshah1992** [comment]
Would also be great to have an itests for this.  

1.) submit transaction -> wait for it to be mined -> fetch trace -> all values should be what we expect
2)  send non-existent tx hash -> fails with the expected error message

`itests/eth_transactions_test.go` shows how this can be done.

File: node/impl/full/eth.go (hunk 1 above)
Comment (snissn) [reply node/impl/full/eth.go:1002]: changed here a46b3d71a

**Comment by snissn** [comment]
> Would also be great to have itests for this. All of the required harnesses are already in place.
> 
> 1.) submit transaction -> wait for it to be mined -> fetch trace -> all values should be what we expect 2) send non-existent tx hash -> fails with the expected error message
> 
> `itests/eth_transactions_test.go` shows how this can be done.

Thank you! Updated the itest file here b3b605fe9

**Review by aarshkshah1992 — APPROVED** [review]
LGTM

**Pull request merged**